canonical/operator
# fix pebble idempotency behavior in testing

This PR changes the behavior of pebble in ops.testing module, to match the behavior of a real pebble server.
The previous code does not allow idempotent operations (start on started service, stop on stopped service) where a real pebble server is idempotent (https://juju.is/docs/sdk/interact-with-pebble#heading--start-and-stop).

Code for checking for previous service status in pebble has been removed in both `start_services` and `stop_services`, which does not match the idempotent behavior of a real pebble server. 
Pebble server does the following as validation which seems unnecessary to be implemented in testing module as of now.
1. Validate names of services
2. Validates dependencies (service before, after) and checks for loops
3. Applies changes with timeout

It should be noted that these services are supposed to be long-running (processes exiting within 1 second will still return pebble.ChangError), I would be happy to add this behavior as a test in `TestRealPebble` under `test/test_pebble.py` `class TestRealPebble`.

## Checklist


       We want to avoid simple passthrough of model data, without contextualization, where possible.


## QA steps

*Please replace with how we can verify that the change works.*

```sh
tox -e unit -- -k test_start_started_service
tox -e unit -- -k test_stop_stopped_service
```

## Documentation changes

No docs needs updating(https://ops.readthedocs.io/en/latest/index.html?highlight=testing#ops.pebble.Client.start_services is fine)

## Bug reference

No bug ref

## Changelog

*Please include a bulleted list of items here, suitable for inclusion in a release changelog.*

- *Fixes testing behavior of start_services and stop_services to be idempotent*
